src/sema/ast.rs

```diff
@@ -228,8 +228,10 @@ pub struct Parameter {
     pub indexed: bool,
     /// Some builtin structs have readonly fields
     pub readonly: bool,
-    /// A struct may contain itself which make the struct infinite size in
-    /// memory. This boolean specifies which field introduces the recursion.
+    /// A recursive struct may contain itself which make the struct infinite size in memory.
+    pub infinite_size: bool,
+    /// Is this struct field recursive. Recursive does not mean infinite size in all cases:
+    /// `struct S { S[] s }` is recursive but not of infinite size.
```

Comment:
I don't get it. Why would implementing it in solidity fix it?

However, it is a good discovery. Recursive types are not allowed in public interfaces anyways and it looks like solc forbids encoding / decoding them:

```solidity
contract Test {
        struct S { int foo; S[] s; }

        function test(bytes memory b) public {
                S memory bar = S({foo: 0, s: new S[](0)});
                abi.encode(bar);
                S memory s = abi.decode(b, (S));
        }
}
```

```
solc --bin /tmp/test.sol

Error: This type cannot be encoded.
 --> /tmp/test.sol:9:14:
  |
9 |             abi.encode(bar);
  |                        ^^^

Error: Decoding type struct Test.S memory not supported.
  --> /tmp/test.sol:10:31:
   |
10 |            S memory s = abi.decode(b, (S));
   |              
```

I'll open another PR once we have this one merged

Reply:
> I don't get it. Why would implementing it in solidity fix it?

In a struct like:
```solidity
struct S {
    int foo;
    S[] s;
}
```

The encoder we have in our compiler would recursively (and indefinitely) be building a for-loop to encode all the vectors within the struct. If we could do this using solidity, we would only enter the recursion if `s.length > 0`. During compile time, we have no information about the size of the vector.

Reply:
I still don't get it, feel free to enlighten me. Everything written in Solidity ends up in our codegen graph, so why is it not possible to write the same code into our CFG manually?

Reply:
I think you're right. If, instead of generating code recursively, we used function calls for encoding arrays after that we have check for their size, we would have no problem.